File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        //Optimal Sol by perifx and suffix calculation
        int n=nums.size();
        int ans=INT_MIN;
        int pref=1;
        int suf=1;
        for(int i=0;i<n;i++){
            if(pref==0)pref=1;
            if(suf==0)suf=1;
            pref*=nums[i];
            suf*=nums[n-i-1];
            ans=max(ans,max(pref,suf));
        }
        return ans;

        //Brute Force O(n2)
        /*
        int n=nums.size();
        int ans=INT_MIN;
        for(int i=0;i<n;i++){
            int currpro=1;
            for(int j=i;j<n;j++){
                currpro*=nums[j];
                ans=max(ans,currpro);
            }
        }
        return ans;
        */
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        ans = -10**18
        prod = 1
        for i in range(n):
            prod*=nums[i]
            ans = max(ans,prod)
            if prod==0:
                prod=1
        prod = 1
        for i in range(n-1,-1,-1):
            prod*=nums[i]
            ans = max(ans,prod)
            if prod==0:
                prod = 1
        return ans 
